src/pip/_internal/operations/install/wheel.py

```diff
@@ -639,6 +641,14 @@ def _generate_file(path, **kwargs):
         installer_file.write(b'pip\n')
     generated.append(installer_path)
 
+    # Document source package hash
+    hash_file_path = os.path.join(dest_info_dir, 'HASH')
```

Comment:
This would need a PEP to standardize this metadata. (see https://www.python.org/dev/peps/pep-0610/ for an example of a similar PEP proposing to add a new file into the dist_info directory).

Reply:
I wrote a pep specifier, and in the following days I will publish it in https://discuss.python.org/c/packaging  
You can see it at https://github.com/NoahGorny/peps/blob/hash-wheel-source/pep-9999.rst (it got messed up for some reason)

Still need to figure out what to do with VCS installs, as we clone entire repos.

Reply:
@NoahGorny  for VCS I'd suggest doing something like a hash of the VCS URL (e.g git+ssh://...) and then adding the VCS revision information (e.g SVN Revision Number, Git Hash, etc). CVS would be more complicated since there's no single value that defines the whole tree, but this could be implemented for each VCS system to make something sensible based on how it operates. $0.02